client/keys/add.go

```diff
@@ -250,92 +255,65 @@ func runAddCmd(cmd *cobra.Command, args []string) error {
 		}
 	}
 
-	info, err := kb.Derive(name, mnemonic, bip39Passphrase, encryptPassword, *bip44Params)
+	//////////////////////////////
+	//////////////////////////////
+
+	info, err := kb.CreateAccount(name, mnemonic, defaultBIP39Passphrase, encryptPassword, account, index)
 	if err != nil {
 		return err
 	}
-	printCreate(info, mnemonic)
-	return nil
-}
-
-func getBIP44ParamsAndPath(path string, flagSet bool) (*hd.BIP44Params, error) {
-	buf := client.BufferStdin()
-	bip44Path := path
-
-	// if it wasn't set in the flag, give it a chance to overide interactively
-	if !flagSet {
-		var err error
-
-		bip44Path, err = client.GetString(fmt.Sprintf("Enter your bip44 path. Default is %s\n", path), buf)
-		if err != nil {
-			return nil, err
-		}
-
-		if len(bip44Path) == 0 {
-			bip44Path = path
-		}
-	}
 
-	bip44params, err := hd.NewParamsFromPath(bip44Path)
-	if err != nil {
-		return nil, err
+	// Recover key from seed passphrase
+	if viper.GetBool(flagRecover) {
+		// Hide mnemonic from output
+		showMnemonic = false
+		mnemonic = ""
 	}
 
-	return bip44params, nil
+	return printCreate(info, showMnemonic, mnemonic)
 }
 
-func printCreate(info keys.Info, seed string) {
+func printCreate(info keys.Info, showMnemonic bool, mnemonic string) error {
 	output := viper.Get(cli.OutputFlag)
+
 	switch output {
 	case "text":
-		fmt.Fprintln(os.Stderr, "")
+		fmt.Println()
 		printKeyInfo(info, Bech32KeyOutput)
 
-		// print seed unless requested not to.
-		if !viper.GetBool(client.FlagUseLedger) && !viper.GetBool(flagNoBackup) {
-			fmt.Fprintln(os.Stderr, "\n**Important** write this seed phrase in a safe place.")
-			fmt.Fprintln(os.Stderr, "It is the only way to recover your account if you ever forget your password.")
-			fmt.Fprintln(os.Stderr)
-			fmt.Fprintln(os.Stderr, seed)
+		// print mnemonic unless requested not to.
+		if showMnemonic {
+			fmt.Printf("\n**Important** write this mnemonic phrase in a safe place.")
+			fmt.Printf("It is the only way to recover your account if you ever forget your password.")
+			fmt.Println()
+			fmt.Println(mnemonic)
 		}
 	case "json":
 		out, err := Bech32KeyOutput(info)
 		if err != nil {
-			panic(err)
+			return err
 		}
-		if !viper.GetBool(flagNoBackup) {
-			out.Seed = seed
+
+		if showMnemonic {
+			out.Mnemonic = mnemonic
 		}
+
 		var jsonString []byte
 		if viper.GetBool(client.FlagIndentResponse) {
 			jsonString, err = cdc.MarshalJSONIndent(out, "", "  ")
 		} else {
 			jsonString, err = cdc.MarshalJSON(out)
 		}
+
 		if err != nil {
-			panic(err) // really shouldn't happen...
+			return err
 		}
-		fmt.Fprintln(os.Stderr, string(jsonString))
+		fmt.Printf(string(jsonString))
 	default:
```

Comment:
As a side note, can't we validate the formats on flag parsing?

Reply:
https://github.com/spf13/viper#establishing-defaults

Reply:
This is actually happening but far far away :) in tendermint. The default value is `text` so the default case will not get triggered.

https://github.com/tendermint/tendermint/blob/dfa9a9a30a666132425b29454e90a472aa579a48/libs/cli/setup.go#L42

https://github.com/tendermint/tendermint/blob/dfa9a9a30a666132425b29454e90a472aa579a48/libs/cli/setup.go#L148

While the validation is happening, given that it is another layer, I would keep this old code there. In case there are changes, this `default` case will provide a safety net there.